store/mongo/datastore_mongo.go

```diff
@@ -603,6 +603,49 @@ func (db *DataStoreMongo) AddAuthSet(ctx context.Context, set model.AuthSet) err
 	return nil
 }
 
+// UpsertAuthSetStatus inserts a new auth set and if it exists, ensures the
+// AuthSet.Status matches the one in authSet.
+func (db *DataStoreMongo) UpsertAuthSetStatus(ctx context.Context, authSet *model.AuthSet) error {
+	c := db.client.Database(DbName).Collection(DbAuthSetColl)
+
+	if authSet.Id == "" {
+		uid := oid.NewUUIDv4()
+		authSet.Id = uid.String()
+	}
+	id := identity.FromContext(ctx)
+	tenantId := ""
+	if id != nil {
+		tenantId = id.Tenant
+	}
+	authSet.TenantID = tenantId
+	if authSet.Timestamp == nil {
+		ts := time.Now()
+		authSet.Timestamp = &ts
+	}
+
+	fltr := bson.D{
+		{Key: dbFieldTenantID, Value: tenantId},
+		{Key: dbFieldIDDataSha, Value: authSet.IdDataSha256},
+		{Key: dbFieldPubKey, Value: authSet.PubKey},
+	}
+	setOnInsert := *authSet
+	setOnInsert.Status = ""
+	if _, err := c.UpdateOne(ctx, fltr, bson.D{
+		{Key: "$setOnInsert", Value: setOnInsert},
+		{Key: "$set", Value: bson.D{
+			{Key: dbFieldStatus, Value: authSet.Status},
+		}},
+	}, mopts.Update().
+		SetUpsert(true)); err != nil {
+		if mongo.IsDuplicateKeyError(err) {
+			return store.ErrObjectExists
```

Comment:
just to double check what if we are preauthorizing an already preauthorized authset?

Reply:
It will behave the same as doing an update with `{$set: {status: "preauthorized"}}` so the document will not be modified.